### ISSUE TITLE
UMAP algorithm does not work as expected

### ISSUE THREAD

**Issue by stefanhahmann** [haifengl/smile#797]
**Describe the bug**
When running the UMAP algorithm with sample data, e.g. two distinct 3d point clouds and trying to reduce them to 2 dimensions, the UMAP does only contain the embeddings of the larger of the two point clouds, but not the smaller one.

**Expected behavior**
The UMAP algorithm result should contain the embedding results for all input data points

**Code snippet**
https://github.com/haifengl/smile/pull/796

**Additional context**
- JDK21
- Smile version 4.0.0
- Windows

**Comment by kklioss** [comment]
It is because the default spectral layout initialization can be applied only on connected components. I add PCA initialization if there are multiple connected components. Please try the master branch. On the other hand, the existence of multiple connected components implies that a global view of the data cannot be attained with this initialization. In this case, you may want to increase k of knn.

**Comment by stefanhahmann** [comment]
> It is because the default spectral layout initialization can be applied only on connected components. I add PCA initialization if there are multiple connected components. Please try the master branch. On the other hand, the existence of multiple connected components implies that a global view of the data cannot be attained with this initialization. In this case, you may want to increase k of knn.

Thanks for the prompt reply. PCA initialization sounds reasonable. I also saw that in other UMAP implementations.
Also, the remark regarding the k makes complete sense.
I tried to build the master branch and test with, but I get 
![grafik](https://github.com/user-attachments/assets/7ab14d3a-d22e-49ef-8fb7-d0cd3c5367ac)

and indeed it seems that the method is not yet in the master branch: https://github.com/haifengl/smile/blob/master/base/src/main/java/smile/graph/AdjacencyList.java

**Comment by stefanhahmann** [comment]
Thanks for the quick fix. The unit test contained in https://github.com/haifengl/smile/pull/796 works with this modification.
Thanks also for the hint regarding k of knn. It makes complete sense for the artificial example of the unit test.

Are you planning to publish a patch release of smile-core/smile-base?

Unrelated, but maybe interesting: I did a performance comparison between UMAP smile implementation and UMAP tagbio implementation (https://github.com/tag-bio/umap-java). Besides the fact that both implementation get slightly different results (which may be due to the random elements inside the UMAP algorithm), the tagbio-implementation seems to be 2-5 times faster, cf.:

* tagbio implementation:
  * https://github.com/mastodon-sc/mastodon-deep-lineage/blob/pca/src/test/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/UmapDemo.java
  * https://github.com/mastodon-sc/mastodon-deep-lineage/blob/pca/src/test/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/UmapDemoIris.java
  * https://github.com/mastodon-sc/mastodon-deep-lineage/blob/pca/src/test/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/UmapDemoTgmmMini.java
* smile implementation:
  *  https://github.com/mastodon-sc/mastodon-deep-lineage/blob/pca/src/test/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/UmapSmileDemo.java
  * https://github.com/mastodon-sc/mastodon-deep-lineage/blob/pca/src/test/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/UmapSmileDemoIris.java
  * https://github.com/mastodon-sc/mastodon-deep-lineage/blob/pca/src/test/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/UmapSmileDemoTgmmMini.java

**Comment by kklioss** [comment]
Have you checked if we use the same number of iterations for both implementations?

**Comment by kklioss** [comment]
A major time in smile is spent on spectral layout as initial layout as it requires eigen value decomposition. It looks like that tagbio doesn't support [spectral](https://github.com/tag-bio/umap-java/blob/9816ba08b6901e74cd9b6df6cc649f56f70cf79b/src/main/java/tagbio/umap/Umap.java#L578) initialization. This may explain the difference on speed and final embedding.

**Comment by stefanhahmann** [comment]
> Significant time of Smile's UMAP is spent on spectral layout for initialization as it requires eigen value decomposition. It looks like that tagbio doesn't support [spectral](https://github.com/tag-bio/umap-java/blob/9816ba08b6901e74cd9b6df6cc649f56f70cf79b/src/main/java/tagbio/umap/Umap.java#L578) initialization. This may explain the difference on speed and final embedding.

Yes, indeed. The tagbio implementation does not support the spectral initialization. It also does not support PCA initialization, which smile now uses, when there are multiple connected components.
And indeed this does shorten the time before doing the actual 500 iterations. The iterations for the optimization seem to take about the same time for both implementations, but the initialization takes longer for the smile implementation, which can be easily understood, since the tagbio initialization uses just a random initilization approach.
In

**Comment by stefanhahmann** [comment]
I did some further profiling. As done in https://github.com/haifengl/smile/blob/master/core/src/test/java/smile/manifold/UMAPTest.java I did some tests with the MNIST dataset, i.e. I run `UMAP umap = UMAP.of(MNIST.x, 7);` and compared it to equivalent tests with the tagbio implementation and the UMAP python implementation.

* It took 18:36 minutes in total
* 18:10 minutes of this were used for initializing the graph `NearestNeighborGraph cc = NearestNeighborGraph.largest(graph);`
* The actual optimization of the layout took about 25s


* The UMAP Python implementation took ~60s and came close to what is stated in the docs (https://umap-learn.readthedocs.io/en/latest/reproducibility.html). 
![python_60s](https://github.com/user-attachments/assets/aee72171-4a0b-43d8-90c1-f242e420601d)
* The tagbio implementation takes ~45s and reaches a qualitatively similar (but numerically different) result: 
![tagbio_45s](https://github.com/user-attachments/assets/e870ca11-8f7b-4554-ae26-77de68151c66)
* The results of the smile implementation seem to differ substantially: 
![smile_1116s](https://github.com/user-attachments/assets/2f99d2c3-b2c8-4e73-83cc-75d4faa5ba88)

**Comment by kklioss** [comment]
Can you please try to increase JVM memory like -Xmx=16G?

**Comment by stefanhahmann** [comment]
I pulled the latest commits (https://github.com/haifengl/smile/commit/82895263d889bc4f5652302d64dd8cb54ba6c81a), re-built, and tried with -Xmx=16G, but it is still slow with the mnist_70000 dataset: 
[mnist_70000.zip](https://github.com/user-attachments/files/18125180/mnist_70000.zip)

The overall performance with the 2500 mnist dataset has now decreased from 4.6s to 15.1s on my machine.

**Comment by kklioss** [comment]
We have added a series of optimization so that UMAP can finish in 4 minutes with full size MNIST data on my laptop. We will look into the difference of embedding in the next.

**Comment by kklioss** [comment]
This is the latest embedding results on full size MNIST data. It takes less than 3 minutes on my PC.
![umap](https://github.com/user-attachments/assets/9abbc855-b7cc-4aa1-919c-dc1e31c1f4be)

**Comment by stefanhahmann** [comment]
> This is the latest embedding results on full size MNIST data. It takes less than 3 minutes on my PC.

The embedding looks a lot better than before. However, I cannot reproduce it with the MINIST 70_000 dataset. Which initialization did you use?


I used:
`
UMAP.of( data, 15);
`
and get this result:
![grafik](https://github.com/user-attachments/assets/3331dbe8-9164-4bbe-9e66-139812df53e8)

The performance improvement is also good to read! It is a bit of a pity that it is still 3-4 times slower than the original python implementation and the other existing java implementation.

**Comment by kklioss** [comment]
Can you please try v4 branch? 

```
double[][] x = Read.csv("mnist_70000.csv").toArray();
long start = System.currentTimeMillis();
var map = UMAP.of(x, 15);
long end = System.currentTimeMillis();
System.out.format("UMAP takes %.2f seconds\n", (end - start) / 1000.0);
ScatterPlot.of(map, '@', Color.ORANGE).canvas().window();
```

The overall process takes 210s on my machine. UMAP algorithm itself takes about 60s, which is on par with python implementation. The rest of time is on nearest neighbor graph construction. I don't know if python implementation includes this part of time into their claim.

**Comment by kklioss** [comment]
Python implementation uses single precision float data type while Smile uses double. I think that it is a major source of performance difference. Java doesn't have c++ like templates. It is hard for us to support both data types with unified source code.

**Comment by haifengl** [comment]
This is included in release v4.1
